gantry/routes/prediction/prediction.py

```diff
@@ -124,6 +127,41 @@ async def get_sample(db: aiosqlite.Connection, build: dict) -> list:
     return []
 
 
+def filter_variants(sample: list, build: dict) -> list:
+    """
+    Filter the sample to match the build's variants.
+    """
+
+    EXPENSIVE_VARIANTS = {
+        "sycl",
+        "mpi",
+        "rocm",
+        "cuda",
+        "python",
+        "fortran",
+        "openmp",
+        "hdf5",
+    }
+
+    matched_rows = []
+    build_variants = spec.spec_variants(build["pkg_variants"])
+
+    # we prefer exact matches but
+    # we can accept the row if all the values of
+    # EXPENSIVE_VARIANTS match
+    for row in sample:
+        row_variants = spec.spec_variants(row[0])
+        if build_variants == row_variants:
+            matched_rows.append(row)
+        elif all(
+            row_variants.get(variant) == build_variants.get(variant)
+            for variant in EXPENSIVE_VARIANTS
+        ):
+            matched_rows.append(row)
```

Comment:
I'd appreciate some advice on how to incorporate this logic into the sql query. With this setup, we can run into a situation where the db returns the five most recent builds, but none of them match the variants we would like to. For instance, if the 5 most recent builds of Trilinos are all CUDA builds, a request for a ROCm prediction would yield no results, giving the client the default prediction, which is not what we want.

Ideally, we would be able to filter for variants directly within the database, but I'll have to look into the best way to efficiently do this. If anyone has tips or thoughts, that would be great!

Reply:
see https://github.com/spack/spack-gantry/pull/5/commits/c52998ff6f063b5969f306c3ad82ff7251432918